pkg/networking/cilium/reconciler/reconciler.go

```diff
@@ -203,6 +205,16 @@ func (r *Reconciler) upgrade(ctx context.Context, logger logr.Logger, client cli
 	}
 
 	logger.Info("Applying Cilium upgrade manifest")
+
+	// When upgrading from Cilium v1.11.x --> Cilium v1.12.x, the port number for a few components changed but the field names remained the same.
+	// This caused "duplicate value" errors when upgrading to the new Cilium version using server-side apply because the merge key for these fields is the port number, not the name.
+	// To alleviate this issue, we will use the client.MergeFrom strategy to completely replace the old object with the new object that is compatible with the new version of Cilium.
+	// We are only doing this for the Cilium Service, DaemonSet, and Deployment since those are the only objects affected.
+	// The rest of the objects in the Cilium upgrade manifest will continue to be applied using server-side apply.
+	if err := r.reconcileSpecialCases(ctx, client, upgradeManifest); err != nil {
+		return controller.Result{}, err
+	}
+
 	if err := serverside.ReconcileYaml(ctx, client, upgradeManifest); err != nil {
 		return controller.Result{}, err
 	}
```

Comment:
This is doing the un-marshaling twice
I would first unmarshall, separate the objects into two slices based on the Kind and then reconcile separately.